alf/networks/actor_distribution_networks.py

```diff
@@ -0,0 +1,165 @@
+from alf.tensor_specs import BoundedTensorSpec, TensorSpec
+
+
+@gin.configurable
+class ActorDistributionNetwork(nn.Module):
+    """Outputs temporally correlated actions."""
+
+    def __init__(self,
+                 input_tensor_spec,
+                 action_spec,
+                 conv_layer_params=None,
+                 fc_layer_params=None,
+                 activation=torch.relu,
+                 discrete_projection_net_ctor=CategoricalProjectionNetwork,
+                 continuous_projection_net_ctor=NormalProjectionNetwork):
+        """Creates an instance of `ActorDistributionNetwork`.
+
+        Args:
+            input_tensor_spec (TensorSpec): the tensor spec of the input
+            action_spec (TensorSpec): the action spec
+            conv_layer_params (list[tuple]): a list of tuples where each
+                tuple takes a format `(filters, kernel_size, strides, padding)`,
+                where `padding` is optional.
+            fc_layer_params (list[int]): a list of integers representing hidden
+                FC layer sizes.
+            activation (nn.functional): activation used for hidden layers.
+            discrete_projection_net_ctor (ProjectionNetwork): constructor that
+                generates a discrete projection network that outputs discrete
+                actions.
+            continuous_projection_net_ctor (ProjectionNetwork): constructor that
+                generates a continuous projection network that outputs
+                continuous actions.
+        """
+        super(ActorDistributionNetwork, self).__init__()
+        self._encoding_net = EncodingNetwork(
+            input_tensor_spec, conv_layer_params, fc_layer_params, activation)
+
+        def _create_projection_net(input_size):
+            if action_spec.is_discrete:
+                assert isinstance(action_spec, BoundedTensorSpec), \
+                    "The action spec of discrete actions must be bounded!"
+                self._projection_net = discrete_projection_net_ctor(
+                    input_size=input_size,
+                    num_actions=action_spec.maximum - action_spec.minimum + 1)
+            else:
+                self._projection_net = continuous_projection_net_ctor(
+                    input_size=input_size, action_spec=action_spec)
+
+        self._create_projection_net = _create_projection_net
+        self._create_projection_net(self._encoding_net.output_size)
+
+    def forward(self, observation, state=()):
+        """Computes an action distribution given an observation.
+
+        Args:
+            observation (torch.Tensor): consistent with `input_tensor_spec`
+            state: empty for API consistent with ActorRNNDistributionNetwork
+
+        Returns:
+            act_dist (torch.distributions): action distribution
+            state: empty
+        """
+        act_dist = self._projection_net(self._encoding_net(observation))
+        return act_dist, state
```

Comment:
need to implement state_spec. Perhaps we should have a common base class Network which provides all the necessary interfaces.